esp-hal-common/src/embassy/time_driver_systimer.rs

```diff
@@ -0,0 +1,117 @@
+use critical_section::{CriticalSection, Mutex};
+
+use super::AlarmState;
+use crate::{
+    clock::Clocks,
+    pac,
+    systimer::{Alarm, SystemTimer, Target},
+};
+
+pub const ALARM_COUNT: usize = 3;
+
+pub struct EmbassyTimer {
+    pub alarms: Mutex<[AlarmState; ALARM_COUNT]>,
+    pub alarm0: Alarm<Target, 0>,
+    pub alarm1: Alarm<Target, 1>,
+    pub alarm2: Alarm<Target, 2>,
+}
+
+const ALARM_STATE_NONE: AlarmState = AlarmState::new();
+
+embassy_time::time_driver_impl!(static DRIVER: EmbassyTimer = EmbassyTimer {
+    alarms: Mutex::new([ALARM_STATE_NONE; ALARM_COUNT]),
+    alarm0: unsafe { Alarm::<_, 0>::conjure() },
+    alarm1: unsafe { Alarm::<_, 1>::conjure() },
+    alarm2: unsafe { Alarm::<_, 2>::conjure() },
+});
+
+impl EmbassyTimer {
+    pub(crate) fn now() -> u64 {
+        SystemTimer::now()
+    }
+
+    pub(crate) fn trigger_alarm(&self, n: usize, cs: CriticalSection) {
+        let alarm = &self.alarms.borrow(cs)[n];
+        // safety:
+        // - we can ignore the possiblity of `f` being unset (null) because of the
+        //   safety contract of `allocate_alarm`.
+        // - other than that we only store valid function pointers into alarm.callback
+        let f: fn(*mut ()) = unsafe { core::mem::transmute(alarm.callback.get()) };
+        f(alarm.ctx.get());
+    }
+
+    fn on_interrupt(&self, id: u8) {
+        match id {
+            0 => self.alarm0.clear_interrupt(),
+            1 => self.alarm1.clear_interrupt(),
+            2 => self.alarm2.clear_interrupt(),
+            _ => unreachable!(),
+        };
+        critical_section::with(|cs| {
+            self.trigger_alarm(id as usize, cs);
+        })
+    }
+
+    pub(crate) fn init(_clocks: &Clocks) {
+        use crate::{interrupt, interrupt::Priority, macros::interrupt};
+
+        // TODO these priorities should probably be higher than 1...
```

Comment:
Are we able to determine a suitable priority for these now, or do we need some time to experiment? Not a big deal if this gets pushed off to another PR.

Reply:
The hardest part is knowing a prio that will work for both riscv and xtensa, maybe ill add a method like `Priority::max()` or something like that.